embassy-usb/src/host/descriptor.rs

```diff
@@ -0,0 +1,591 @@
+                })
+            {
+                self.buffer_idx += EndpointDescriptor::SIZE;
+                self.index += 1;
+                return Some(endpoint);
+            }
+            None
+        }
+    }
+}
+
+/// InterfaceDescriptor does not implement [USBDescriptor] because it has a borrowed buffer.
+/// Since we cannot request an interface decriptor from the device by itself it does not strictly need to implement [USBDescriptor].
+impl<'a> InterfaceDescriptor<'a> {
+    const SIZE: usize = 9;
+
+    const DESC_TYPE: u8 = descriptor_type::INTERFACE;
+
+    fn try_from_bytes(bytes: &'a [u8]) -> Result<Self, ()> {
+        if bytes.len() < Self::SIZE {
+            return Err(());
+        }
+        if bytes[1] != Self::DESC_TYPE {
+            return Err(());
+        }
+        Ok(Self {
+            len: bytes[0],
+            descriptor_type: bytes[1],
+            interface_number: bytes[2],
+            alternate_setting: bytes[3],
+            num_endpoints: bytes[4],
+            interface_class: bytes[5],
+            interface_subclass: bytes[6],
+            interface_protocol: bytes[7],
+            interface_name: bytes[8],
+            buffer: &bytes[Self::SIZE..],
+        })
+    }
+
+    /// Try to parse a class descriptor of a given type
+    pub fn parse_class_descriptor<T: USBDescriptor>(&self) -> Option<T> {
+        Self::identify_descriptor::<T>(self.buffer).and_then(|i| T::try_from_bytes(&self.buffer[i..]).ok())
+    }
+
+    /// Iterate over endpoints
+    pub fn iter_endpoints(&'a self) -> EndpointIterator<'a> {
+        EndpointIterator {
+            index: 0,
+            buffer_idx: 0,
+            iface_desc: self,
+        }
+    }
+
+    /// Parse up to `L` endpoints corresponding to this interface.
+    /// Returns a vector of EndpointDescriptors. The length of the vector is `min(L, self.num_endpoints)`.
+    pub fn parse_endpoints<const L: usize>(&self) -> Vec<EndpointDescriptor, L> {
+        let mut endpoints: Vec<EndpointDescriptor, L> = Vec::new();
+
+        let mut working_buffer = self.buffer;
+        for _ in 0..self.num_endpoints.min(L as u8) {
+            if let Some(endpoint) = Self::identify_descriptor::<EndpointDescriptor>(working_buffer).and_then(|i| {
+                working_buffer = &working_buffer[i..];
+                EndpointDescriptor::try_from_bytes(working_buffer).ok()
+            }) {
+                // safe because we limited the iterations.
+                endpoints.push(endpoint).ok();
+            }
+            working_buffer = &working_buffer[EndpointDescriptor::SIZE..];
+        }
+
+        endpoints
+    }
+
+    // Returns the offset to the first matching descriptor in the slice
+    fn identify_descriptor<T: USBDescriptor>(slice: &[u8]) -> Option<usize> {
+        let mut offset = 0;
+        let mut desc_len = slice[offset] as usize;
+        let mut desc_type = slice[offset + 1];
+
+        while desc_type != T::DESC_TYPE || desc_len != T::SIZE {
```

Comment:
Only accepting descriptors that are the exact size expected as is done in this line would make the descriptor fail to parse if there were existance of any extentions to the descriptors such as  those that exist in USB 2.0+.

For example MIDI 2.0 DeviceEndpoints have can potentially have two extra bytes in positions 7 and 8 and so the DeviceEndpointDescriptor bLength may be 9 rather than 7.

| Offset | Field         | Size | Value | Description |
|--------|---------------|------|-------|-------------|
| 0-6  |     |     |   | (same as in USB 1.1)      |
| 7      | `bRefresh`    | 1    | 0x00  | Unused      |
| 8      | `bSynchAddress` | 1  | 0x00  | Unused      |

What this means is in practice if you assume exact sizes, newer devices will be ignored by this driver if they use any extensions even though they should be backwards compatible.

The assertion that bytes beyond the expected size must be ignored is mentioned briefly in the USB spec in Chapter 9.5 "Descriptors":
> If a descriptor returns with a value in its length field that is less than defined by this specification, the
descriptor is invalid and should be rejected by the host. If the descriptor returns with a value in its length field that is greater than defined by this specification, the extra bytes are ignored by the host, but the next
descriptor is located using the length returned rather than the length expected.

This would also be relevant to `parse_endpoints` (and other descriptor parsing logic above) which assumes the record length when advancing the working_buffer slice `working_buffer = &working_buffer[EndpointDescriptor::SIZE..];` as that could break if some bytes in the extended area happened to collide with a valid start to a descriptor.

I think it would be best to refactor the code to always advance the slice being considered by the amount specified by `bLength` of the current slice, and never assume that it's safe to advance by a static amount.

Reply:
A sent an attempted fix as another PR targeting this PR branch. Please take a look :)